# Author: Стурейко Игорь
## Project: Geekbrains.MatStat
## Lesson 07 - Линейная алгебра и матрицы
## Date: 2020-06-25

## 5.1. Вектор – это частный случай матрицы 1хN и Nх1. Повторите материал для векторов, уделяя особое внимание умножению A∙B. Вычислите, по возможности не используя программирование: (5E)-1 где Е – единичная матрица размера 5х5.

Исходя из определения и свойств единичной и обратной матрицы: пусть $a = 5\cdot E$, тогда $b = (5 \cdot E)^{-1}$ такова, что $a\cdot b = E$, т.е $b = \frac{1}{5}\cdot E$

In [1]:
import numpy as np
E = np.eye(5)
E

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [2]:
a = 5*E
a

array([[5., 0., 0., 0., 0.],
       [0., 5., 0., 0., 0.],
       [0., 0., 5., 0., 0.],
       [0., 0., 0., 5., 0.],
       [0., 0., 0., 0., 5.]])

In [3]:
b = E/5
b

array([[0.2, 0. , 0. , 0. , 0. ],
       [0. , 0.2, 0. , 0. , 0. ],
       [0. , 0. , 0.2, 0. , 0. ],
       [0. , 0. , 0. , 0.2, 0. ],
       [0. , 0. , 0. , 0. , 0.2]])

In [4]:
c = np.dot(a, b)
c

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

## 5.2. Вычислите определитель:

$$\begin{vmatrix}
1 & 2 & 3 \\
4 & 0 & 6 \\
7 & 8 & 9
\end{vmatrix}$$

Поскольку во второй строке есть 0, то раскладывать будем по второй строке:

$$det \begin{vmatrix}
1 & 2 & 3 \\
4 & 0 & 6 \\
7 & 8 & 9
\end{vmatrix} = -4\cdot \begin{vmatrix}
2 & 3 \\
8 & 9
\end{vmatrix} + 0 \cdot \begin{vmatrix}
1 & 3 \\
7 & 9
\end{vmatrix}
- 6 \cdot \begin{vmatrix}
1 & 2 \\
7 & 8
\end{vmatrix} = -4\cdot (2\cdot 9 - 3 \cdot 8) - 6\cdot (1\cdot 8-2 \cdot 7)=60$$

In [5]:
a = [[1, 2, 3], [4, 0, 6], [7, 8, 9]]
a

[[1, 2, 3], [4, 0, 6], [7, 8, 9]]

In [6]:
np.linalg.det(a)

59.999999999999986

## 5.3.1 Вычислите матрицу, обратную данной:

$$B = \begin{vmatrix}
1 & 2 & 3 \\
4 & 0 & 6 \\
7 & 8 & 9
\end{vmatrix}$$

1. Найдем матрицу миноров $M$

$$M = \begin{vmatrix}
-48 & -6 & 32 \\
-6 & -12 & -6 \\
12 & -6 & -8
\end{vmatrix}$$

2. Находим матрицу алгебраических дополнений $A$

$$A = \begin{vmatrix}
-48 & 6 & 32 \\
6 & -12 & 6 \\
12 & 6 & -8
\end{vmatrix}$$

3. Транспонируем ее:

$$A^T= \begin{vmatrix}
-48 & 6 & 12 \\
6 & -12 & 6 \\
32 & 6 & -8
\end{vmatrix}$$

4. Формируем обратную:

$$B^{-1}=\frac{1}{det{B}}\cdot A^T=\frac{1}{60}\cdot \begin{vmatrix}
-48 & 6 & 12 \\
6 & -12 & 6 \\
32 & 6 & -8
\end{vmatrix}$$

In [7]:
a = [[1, 2, 3], [4, 0, 6], [7, 8, 9]]
b = np.linalg.inv(a)
b

array([[-0.8       ,  0.1       ,  0.2       ],
       [ 0.1       , -0.2       ,  0.1       ],
       [ 0.53333333,  0.1       , -0.13333333]])

In [8]:
np.linalg.det(a)

59.999999999999986

## 5.3.2 Приведите пример матрицы 4х4, ранг которой равен 1.

Все строки представляют собой линейно зависимые

In [9]:
a = np.matrix([[1, 2, 3, 4], [2, 4, 6, 8], [3, 6, 9, 12], [10, 20, 30, 40]])
a, np.linalg.matrix_rank(a)

(matrix([[ 1,  2,  3,  4],
         [ 2,  4,  6,  8],
         [ 3,  6,  9, 12],
         [10, 20, 30, 40]]),
 1)

## 5.4 Вычислите скалярное произведение двух векторов:  (1, 5)  и  (2, 8)

Скалярное произведение возвращает функция dot, скалярное произведение равно: $\vec{a}\cdot \vec{b} = a\cdot b \cdot \cos{\alpha}$

In [10]:
a = np.array([1,5])
b = np.array([2,8])
np.dot(a, b)

42

## 5.5 Вычислите смешанное произведение трех векторов:(1, 5, 0), (2, 8, 7) и (7, 1.5, 3) 

In [11]:
a = np.array([1, 5, 0])
b = np.array([2, 8, 7])
c = np.array([7, 1.5, 3])


Смешанное произведение $\vec{a}\cdot(\vec{b}\times\vec{c})$, физический смысл - объем параллепипеда

In [12]:
np.dot(a, (np.cross(b, c)))

228.5

## 6.1 Решите линейную систему:

$$\begin{bmatrix}
1 & 2 & 3 \\
4 & 0 & 6 \\
7 & 8 & 9
\end{bmatrix} \cdot X = 
\begin{bmatrix}
12 \\ 2 \\ 1
\end{bmatrix}
$$

In [13]:
a = np.matrix([[1, 2, 3], [4, 0, 6], [7, 8, 9]])
b = np.matrix([12, 2, 1]).T
a, b

(matrix([[1, 2, 3],
         [4, 0, 6],
         [7, 8, 9]]),
 matrix([[12],
         [ 2],
         [ 1]]))

In [14]:
np.linalg.solve(a, b)

matrix([[-9.2       ],
        [ 0.9       ],
        [ 6.46666667]])

## 6.2 Найдите псевдорешение:

$$\begin{cases}
x+2y-z=1\\
3x-4y=7\\
8x-5y+2z=12\\
2x-5z=7\\
11x+4y-7z=15\end{cases}
$$

In [15]:
a = np.matrix([[1, 2, -1],[3, -4, 0],[8, -5, 2],[2, 0, -5],[11, 4, -7]])
b = np.matrix([1, 7, 12, 7, 15]).T
a, b

(matrix([[ 1,  2, -1],
         [ 3, -4,  0],
         [ 8, -5,  2],
         [ 2,  0, -5],
         [11,  4, -7]]),
 matrix([[ 1],
         [ 7],
         [12],
         [ 7],
         [15]]))

In [16]:
np.linalg.lstsq(a, b)

/Users/igor/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  """Entry point for launching an IPython kernel.


(matrix([[ 1.13919353],
         [-0.90498444],
         [-0.9009803 ]]),
 matrix([[0.71523211]]),
 3,
 array([15.2817306 ,  9.59852942,  3.65197794]))

## 6.3 Сколько решений имеет линейная система:

$$\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{bmatrix} \cdot X = 
\begin{bmatrix}
12 \\ 2 \\ 1
\end{bmatrix}
$$

Если ноль – то измените вектор правой части так, чтобы система стала совместной, и решите ее.

In [17]:
a=np.matrix([[1, 2, 3], [4, 5, 6],[7, 8, 9]])
b = np.matrix([12,2,-8])
c = np.concatenate((a, b.T), axis=1)
c, np.linalg.matrix_rank(a, 0.001), np.linalg.matrix_rank(c, 0.001)

(matrix([[ 1,  2,  3, 12],
         [ 4,  5,  6,  2],
         [ 7,  8,  9, -8]]),
 2,
 2)

## 6.4 Вычислите LU-разложение матрицы:

$$\begin{bmatrix}
1 & 2 & 3 \\
2 & 16 & 21 \\
4 & 28 & 73
\end{bmatrix}$$

После этого придумайте вектор правых частей и решите полученную линейную систему трех уравнений с данной матрицей.

In [23]:
import numpy as np
import scipy.linalg as la
a = np.array([[1, 2, 3], [2, 16, 21], [4, 28, 73]])
(P, L, U) = la.lu(a)
P, L, U

(array([[0., 1., 0.],
        [0., 0., 1.],
        [1., 0., 0.]]),
 array([[ 1.  ,  0.  ,  0.  ],
        [ 0.25,  1.  ,  0.  ],
        [ 0.5 , -0.4 ,  1.  ]]),
 array([[  4.  ,  28.  ,  73.  ],
        [  0.  ,  -5.  , -15.25],
        [  0.  ,   0.  , -21.6 ]]))

In [24]:
b = np.array([1, 2, 3])
np.linalg.solve(a, b)

array([ 1.01388889,  0.03472222, -0.02777778])

## 6.5 Найдите нормальное псевдорешение недоопределенной системы:

$$\begin{cases}
x+2y-z=1\\
8x-5y+2z=12\end{cases}
$$

Для этого определите функцию Q(x,y,z), равную норме решения, и найдите ее минимум.

In [30]:
a = np.array([[1, 2, -1],[8, -5, 2]])
b = np.array([1, 12])
np.linalg.lstsq(a, b)[0]

/Users/igor/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  This is separate from the ipykernel package so we can avoid doing imports until


array([ 1.38191882, -0.18081181,  0.0202952 ])

Точка нормального псевдорешения (1.38191882, -0.18081181,  0.0202952)

## 6.6 Найдите одно из псевдорешений вырожденной системы:

$$\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{bmatrix} \cdot X = 
\begin{bmatrix}
2 \\ 5 \\ 11
\end{bmatrix}
$$

Попробуйте также отыскать и нормальное псевдорешение.

In [27]:
a = np.array([[1, 2, 3], [4, 5, 6],[7, 8, 9]])
b = np.array([2, 5, 11])
np.linalg.matrix_rank(a, 0.001)

2

In [28]:
np.linalg.lstsq(a, b)

/Users/igor/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  """Entry point for launching an IPython kernel.


(array([ 1.25,  0.5 , -0.25]),
 array([], dtype=float64),
 2,
 array([1.68481034e+01, 1.06836951e+00, 3.33475287e-16]))

Нормальное псевдорешение (1.25, 0.5, -0.25)